cpp/src/c_api/uniform_neighbor_sampling.cpp

```diff
@@ -222,6 +222,62 @@ extern "C" cugraph_type_erased_host_array_view_t* cugraph_sample_result_get_coun
   return reinterpret_cast<cugraph_type_erased_host_array_view_t*>(internal_pointer->count_->view());
 }
 
+extern "C" cugraph_error_code_t cugraph_test_sample_result_create(
+  const cugraph_resource_handle_t* handle,
+  const cugraph_type_erased_device_array_view_t* srcs,
+  const cugraph_type_erased_device_array_view_t* dsts,
+  const cugraph_type_erased_device_array_view_t* weights,
+  const cugraph_type_erased_device_array_view_t* counts,
+  cugraph_sample_result_t** result,
+  cugraph_error_t** error)
+{
+  *result = nullptr;
+  *error  = nullptr;
+  size_t n_bytes{0};
+  cugraph_error_code_t error_code{CUGRAPH_SUCCESS};
+
+  if (!handle) {
+    *error = reinterpret_cast<cugraph_error_t*>(
+      new cugraph::c_api::cugraph_error_t{"invalid resource handle"});
+    return CUGRAPH_INVALID_HANDLE;
+  }
+
+  // copy srcs to new device array
+  cugraph_type_erased_device_array_t* new_device_srcs{nullptr};
+  error_code =
+    cugraph_type_erased_device_array_create_from_view(handle, srcs, &new_device_srcs, error);
+  if (error_code != CUGRAPH_SUCCESS) return error_code;
```

Comment:
This looks much cleaner.

I think this still potentially leaks memory.  Putting the comment here, but probably this one is fine, the rest would be potentially a problem.  If one of the `cugraph_type_erased_device_array_create_from_view` functions fails, then you will return an error code here, but any of the previously allocated device arrays will leak.

Seems like the best option would be to make that an `std::unique_ptr` and define it's freeing function as the `cugraph_type_erased_device_array_free` function (and do that for each of the device array's you create in this function).

Then in the line where we create the return value you can release the unique pointer when calling the constructor.

NOTE: This probably exists in other functions as well, I just noticed it here.

Reply:
Thanks, I added this to my latest commit.